pkg/compute/executor.go

```diff
@@ -86,141 +134,229 @@ func PrepareRunArguments(
+
+	if e.failureInjection.IsBadActor {
+		result.Err = fmt.Errorf("i am a baaad node. i failed execution %s", execution.ID)
+		return
+	}
+
+	if err := jobExecutor.Start(ctx, args); err != nil {
+		log.Ctx(ctx).Error().Err(err).Msg("failed to start execution")
+		result.Err = err
 	}
 
+	return
+}
+
+func (e *BaseExecutor) Wait(ctx context.Context, state store.LocalExecutionState) (*models.RunCommandResult, error) {
+	execution := state.Execution
 	jobExecutor, err := e.executors.Get(ctx, execution.Job.Task().Engine.Type)
 	if err != nil {
-		return fmt.Errorf("failed to get executor %s: %w", execution.Job.Task().Engine, err)
+		return nil, fmt.Errorf("failed to get executor %s: %w", execution.Job.Task().Engine, err)
 	}
 
-	if e.failureInjection.IsBadActor {
-		return fmt.Errorf("i am a baaad node. i failed execution %s", execution.ID)
+	waitC, errC := jobExecutor.Wait(ctx, execution.ID)
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res := <-waitC:
+		return res, nil
+	case err := <-errC:
+		log.Ctx(ctx).Error().Err(err).Msg("failed to wait on execution")
+		return nil, err
 	}
 
-	runCommandCleanup := system.NewCleanupManager()
-	runCommandArguments, err := PrepareRunArguments(ctx, e.Storages, execution, resultFolder, runCommandCleanup)
-	if err != nil {
-		return err
+}
+
+// Run the execution after it has been accepted, and propose a result to the requester to be verified.
+func (e *BaseExecutor) Run(ctx context.Context, state store.LocalExecutionState) (err error) {
+	execution := state.Execution
+	ctx = log.Ctx(ctx).With().
+		Str("job", execution.Job.ID).
+		Str("execution", execution.ID).
+		Logger().WithContext(ctx)
+
+	operation := "Running"
+	defer func() {
+		if err != nil {
+			e.handleFailure(ctx, state, err, operation)
+		}
+	}()
+
+	res := e.Start(ctx, execution)
+	defer func() {
+		if err := res.Cleanup(ctx); err != nil {
+			log.Ctx(ctx).Error().Err(err).Msg("failed to clean up start arguments")
+		}
+	}()
+	if err := res.Err; err != nil {
+		if errors.Is(err, executor.ErrAlreadyStarted) {
+			// by not returning this error to the caller when the execution has already been started/is already running
+			// we allow duplicate calls to `Run` to be idempotent and fall through to the below `Wait` call.
+			log.Ctx(ctx).Warn().Err(err).Str("execution", execution.ID).
+				Msg("execution is already running processing to wait on execution")
+		} else {
+			// We don't consider the job failed if the execution is already running or has already completed.
+			// TODO(forrest): [correctness] do we really want to record a job failed metric if (one of) its execution(s)
+			// failed to start? Perhaps it would be better to have metrics for exection failures here and job failures
+			// higher up the call stack?
+			jobsFailed.Add(ctx, 1)
+			return err
+		}
 	}
-	defer runCommandCleanup.Cleanup(ctx)
 
-	runCommandResult, err := jobExecutor.Run(ctx, runCommandArguments)
+	result, err := e.Wait(ctx, state)
```

Comment:
Given that we no longer call `jobExectuor.Run`, do we need executor plugins to implement a `Run` method anymore? Can we remove it from the interface?

Reply:
It's still used extensively in our executor scenario test cases as the method offers some conveniences for running an execution and returning its output. Would be be okay to punt the removal of this from the interface into a follow on when we are more prepared for the full async nature of the new executor interface?

Reply:
Sure thing!